Messaging service

tristan957 opened the pull request:
Abstract GroupMeSevice further into an impl of MessagingService in preparation for a potential re-org of the code.

tristan957 committed:
Add fp-ts as a dependency

tristan957 committed:
Rework how GroupMeService to a MessagingService impl

JosephMart commented on src/handlers/HeyHandler.ts:
```diff
@@ -11,9 +13,9 @@ export default class HeyHandler implements Handler {
   };
 
-  constructor(@inject(GroupMeService) private readonly groupMeService: GroupMeService) {}
+  constructor(@inject("MessagingService") private readonly messagingService: MessagingService) {}
```
Why is this a string now?

tristan957 replied:
Because it is an interface. Classes have prototypes for getting injection tokens, but interfaces do not.

tristan957 replied:
TS being held back by JS is what I think it boils down to.

JosephMart reviewed (approved):
Not sure I understand the string in the injection part enough but LGTM :)

JosephMart committed:
Merge branch 'master' into messaging-service

tristan957 commented:
#10 should be merged into this PR before merging to master.

JosephMart committed:
Merge branch 'master' into messaging-service

JosephMart reviewed (approved):
assuming it runs :P, LGTM

tristan957 committed:
Move HandlerRegistry to RootHandler (#10)

* Move HandlerRegistry to RootHandler

This was discussed over GroupMe, but handlers should only have
one regex expression. Each Handler then has sub Handlers that can do
further matching.

* Use flat

Co-Authored-By: Joseph Martinsen <joseph@martinsen.com>

* Add documentation around the new structure of Handlers

* Make handler info-agnostic

Handler as an abstract class should be able to handle different types of information
depending on the service that it is implemented for.

* Update src/handlers/HeyHandler.ts

Co-Authored-By: Joseph Martinsen <joseph@martinsen.com>

Co-authored-by: Joseph Martinsen <joseph@martinsen.com>
